ENT-978 Add user country code and last activity date in enterprise en…

zubair-arbi opened the pull request:
…rollment hive table

* Add fields `user_country_code` and `last_activity_date` in `EnterpriseEnrollmentRecord` table.
* Update `insert_query`to populate above fields during `EnterpriseEnrollmentDataTask`

Remaining Task:
* Cannot add/populate field `country_name` as we are using django-countries which provide method for getting country name against the provided country code. We cannot do this with such ease in SQL. One way we can try is using manual dump for countries along with code and name.

georgebabey commented on edx/analytics/tasks/enterprise/enterprise_enrollments.py:
```diff
@@ -127,10 +129,12 @@ def insert_query(self):
                     END AS course_duration_weeks,
                     course.min_effort AS course_min_effort,
                     course.max_effort AS course_max_effort,
+                    course.catalog_course AS course_key,
```
Order matters here so be careful moving course_key. This order here needs to match the order in the fixture data here or else the acceptance tests will fail: edx/analytics/tasks/tests/acceptance/test_enterprise_enrollments.py

See this PR from @mattdrayer https://github.com/edx/edx-analytics-pipeline/commit/16780bdba521035326d492d29cd156a8e843f0e9

georgebabey commented:
Just realized we should call this `last_activity_timestamp` to align with the other datetime fields: `enrollment_created_timestamp`,  `passed_timestamp`, `user_account_creation_timestamp`

brianhw commented:
A few observations:
* You have added references to two additional tables in your Hive query, but you don't do what is needed to load that data into Hive.   In particular, the EnterpriseEnrollmentDataTask.requires() method should include tasks that make this happen.  That includes ImportAuthUserProfileTask() (to load auth_userprofile into Hive), and something to load user_activity_by_user into Hive.  
* The loading of user_activity_by_user into Hive is a little subtle.   You would use UserActivityTableTask (as defined in insights/user_activity.py), but you must pass in a parameter value for overwrite_n_days of zero, so that there is no attempt to recalculate any data.  This would simply use whatever data is currently available on the machine.  
* Your join "ON enterprise_user.user_id = user_profile.id" should be "ON enterprise_user.user_id = user_profile.user_id" instead? 
* In testing, you need to define user_activity_by_user fixture differently.   The data is not read from an LMS database, so there is no need to define such a table.  You have to write out files to the right warehouse_path location, since it is an intermediate output of a different pipeline job.   An example of this is CourseBlocksPartitionTaskAcceptanceTest.copy_hive_input() in test_course_blocks.py, though that's just the first I found.    You may want activity for more than one day.
* You can run acceptance tests on the test_ci server, using the ad-hoc manual job and putting in your branch and the name of the test (i.e. test_enterprise_enrollments).  Let me know if you need permission and don't have it.   Brittney and George do as well.

brianhw commented:
I'm glad you tried to run the acceptance test.  It looks like it failed because of an extra character in one of your fixture files.  

Also, the Travis builds are failing due to an issue with ordering of includes:  
```
isort --check-only --recursive edx/
ERROR: /edx/app/analytics_pipeline/analytics_pipeline/edx/analytics/tasks/enterprise/enterprise_enrollments.py Imports are incorrectly sorted. 
```

zubair-arbi commented:
@brianhw Please have another look on the changes now.
Can you please tell why acceptance tests on jenkins failed: http://jenkins-ci.analytics.edx.org/view/ad-hoc/job/edx-analytics-pipeline-acceptance-test-manual/1566/console

rao-abdul-mannan commented:
@zubair-arbi Error seems to be 
`Failed to recognize predicate 'date'. Failed rule: 'identifier' in table or column identifier`
in the EnterpriseEnrollmentDataTask, which could be due to no whitespace being used in table name and alias `social_auth_usersocialauth`
https://github.com/edx/edx-analytics-pipeline/blob/3f3de316aae718348fc991343805dbf7c7c795c3/edx/analytics/tasks/enterprise/enterprise_enrollments.py#L176

rao-abdul-mannan commented:
Sorry about my earlier comment. 
So I have found 2 issues that **could** be the failure reason.
-  Due to hive reserved keywords, table & column names in the query should be wrapped in back ticks ( ` )

    e.g. date should be `date` in subquery
- `last_activity_timestamp` is a timestamp field and query is writing Date to it.

brianhw commented on edx/analytics/tasks/enterprise/enterprise_enrollments.py:
```diff
@@ -50,6 +51,8 @@ class EnterpriseEnrollmentRecord(Record):
     user_username = StringField(length=255, description='')
     course_key = StringField(length=255, description='')
+    user_country_code = StringField(length=2, description='')
+    last_activity_timestamp = DateTimeField(description='')
```
I think that if you are storing values like "2018-05-18", you may need to use a DateField here instead of a DateTimeField.    I don't know if that would involve changing the name of the field as well.   Or if we need to find a different way of getting the actual timestamp of the last activity, instead of merely the date.

brianhw commented on edx/analytics/tasks/enterprise/enterprise_enrollments.py:
```diff
@@ -141,6 +146,21 @@ def insert_query(self):
+                    SELECT
+                        user_id,
+                        course_id,
+                        MAX(date) AS latest_date
```
This may be the line that requires date to be back-quoted.

brianhw commented:
Looks like it needs a rebase....

brianhw commented on edx/analytics/tasks/tests/acceptance/test_enterprise_enrollments.py:
```diff
@@ -107,7 +114,8 @@ def validate_enterprise_enrollment_table(self):
                    'course_start', 'course_end', 'course_pacing_type', 'course_duration_weeks', 'course_min_effort',
-                   'course_max_effort', 'user_account_creation_timestamp', 'user_email', 'user_username', 'course_key']
+                   'course_max_effort', 'user_account_creation_timestamp', 'user_email', 'user_username', 'course_key',
+                   'user_country_code', 'last_activity_timestamp']
```
You probably need to update the field name here as well.

zubair-arbi commented on edx/analytics/tasks/tests/acceptance/test_enterprise_enrollments.py:
```diff
@@ -40,6 +40,10 @@ def setUp(self):
             url_path_join(self.warehouse_path, 'discovery_api_raw', 'dt=' + self.DATE, 'programs.json')
         )
+        self.upload_file(
+            os.path.join(self.data_dir, 'input', 'internal_reporting_user_activity'),
```
@brianhw I have verified that users defined in `user_activity_by_user` fixture file matches in `user_ids` and also the `upload_file` path matches the hive path `edx.analytics.tasks.insights.user_activity.UserActivityTask#output_path_for_key`. Still `last_activity_date` values are empty.

brianhw replied:
RIght.   But this line uploads 'internal_reporting_user_activity' to the warehouse_path/'user_activity_by_user' location, rather than the 'user_activity_by_user' that you defined earlier.   You need to change this line to os.path.join(self.data_dir, 'input', 'user_activity_by_user'),

zubair-arbi commented:
@brianhw @rao-abdul-mannan I have updated the file path according to your suggestions now acceptance tests are passing with success http://jenkins-ci.analytics.edx.org/view/ad-hoc/job/edx-analytics-pipeline-acceptance-test-manual/1592/console
I have rebased and squashed the commits. Please have a final look on the changes.

brianhw commented:
I've changed the "enterprise-enrollment" job to run off of origin/release instead of origin/master, so there is less risk to production to have this merge to master.   Same with #512 .

zubair-arbi commented:
@brianhw the acceptance tests again passed with success after rebase http://jenkins-ci.analytics.edx.org/view/ad-hoc/job/edx-analytics-pipeline-acceptance-test-manual/1594/console

zubair-arbi commented:
@saleem-latif I have merged this branch please rebase your branch (https://github.com/edx/edx-analytics-pipeline/pull/512) now.
